### ISSUE TITLE
protection main table - column update

### ISSUE THREAD

**Issue by ashachaf** [bancorprotocol/webapp#426]
please update the main protection table.

https://www.figma.com/file/3yED7o7TgoLwoEjV2EkRLL/Bancor-V2?node-id=736%3A1238
columns:
initial stake - same information as the current "stake" column.  just title update.
protected value - same information as the "fully protected value" column. just title update.
claimable value - pull the same data as protected value but use current date (instead of future date). this will return the current available amount based on current protection level.
fees - TBD. hide this column for now.
ROI - calculation already exists. will not include this column in production at this stage as it is still being worked on.
APR - calculation already exists. 
current coverage - few changes here, please note:
1.  main blue timeline bar - dark blue increase every 60seconds using the calculation out of the maxProtectionDelay. currently, defined at 8640000 seconds. which means, the bar should increase ini 60/8640000 every 1min. dark blue displays the. accumulative time from initial stake.
example:
if stake timeframe is 10:30pm -> at 10:40pm you will fill the bar with 600seconds/8640000
2. red indication for cliff - using the minProtectionDelay value:
- display the red indication only when stake timeframe is shorter than minProotectionDelay
example: stake at 10:30pm & cliff is 600second -> display the red indication for the 600seconds from stake.
- indicate the time left until cliff is reached in xxd xxh xxm xxs format. make sure we update every second.
example: 23d 8h 57h 13s 

Tool tips:

- Initial stake = amount of tokens you originally staked in the pool
- Protected value = amount of tokens you can withdraw with 100% protection + fees
- Claimable value = amount of tokens you can withdraw right now (assuming you have not earned full protection, this value will be lower than Protected Value)
- Fees = fees your stake has earned since you entered the pool
- ROI = the ROI of your fully protected value vs. your initial stake
- APR = how much the pool has earned within different time frames
- Cliff: Until the "cliff" is reached, you are entitled to exercise 0% of your impermanent loss protection. When the cliff is reached, you are entitled to 30% coverage. Coverage increases by 1% per day until 100% coverage is reached (full protection).

**Comment by ashachaf** [comment]
position fee calculation:

Fee calculation process:
from the original staked position  you can extract:
user address
pool token address
reserve token address
original pool token amount
original reserve token amount
rate0: 2nd/1st conversion rate
timestamp
calculate the current amount of the reserve token
tkn amount1 or bnt amount1 (depending on the reserve used)  = 
original pool token amount * (current reserve balance / current pool token balance)

calculate the current tkn/bnt rate
rate1 = current 2nd reserve balance / current 1st reserve balance

now we can calculate the fees percentage that are collected on the stake position:

calculate fee percentage
if bnt: (bnt amount1 / bnt amount0)*(rate1 / rate0)^0.5 - 1
if tkn: (tkn amount1 / tkn amount0)*(rate1 / rate0)^0.5 - 1
calculate fee amount in reserve token:
fee amount = original reserve token amount * fee percentage

**Comment by Velua** [comment]
@pingustar can you make sure we're tooltip ready? e.g. can add these tooltips to the columns array or something like that.

**Comment by Velua** [comment]
Tasks I'll take from this 

## Update protected positions getter
- [ ] Offer Claimable value column and add to table.
- [ ] Offer fees column in dec percentage form.

**Comment by pingustar** [comment]
@Velua I have added the tooltips and updated the column titles.

Whats the status for the fee and claimable column, and cliff. Is there something I can implement yet?

**Comment by Velua** [comment]
Cliff is already included as ```insuranceStart```, I've added Claimable value in that drafted PR. 

If you've done those then I imagine it's just waiting on me to deliver the fee column

**Comment by ashachaf** [comment]
@Velua @pingustar 
we are expecting a contract update.
do not push to production until we handle contract update.

**Comment by ashachaf** [comment]
@Velua @pingustar 
please make sure everything is ready from your end.
new contract and new api will be ready next few days and will require you passing an additional parameter to the existing call (same response).

**Comment by pingustar** [comment]
>     1. main blue timeline bar - dark blue increase every 60seconds using the calculation out of the maxProtectionDelay. currently, defined at 8640000 seconds. which means, the bar should increase ini 60/8640000 every 1min. dark blue displays the. accumulative time from initial stake.
>        example:
>        if stake timeframe is 10:30pm -> at 10:40pm you will fill the bar with 600seconds/8640000

@ashachaf is updating the blue timelinebar every 60 sec necessary? I dont think users will notice the progress change given 3 month broken down to 60s steps is probably not a visible change.

**Comment by ashachaf** [comment]
do you suggest to update it once a day?

**Comment by pingustar** [comment]
Currently it updates the progress bar with every visit. So if you refresh you always get the updated status.

**Comment by ashachaf** [comment]
yes, progress should include cliff time.
which means even if i am on day 10 out of 100, and the cliff is 30 days, i should see 10% blue bar.

**Comment by pingustar** [comment]
@Velua I have updated the `calculateProtectionLevel` function in commit 72ef000 to include the cliff time in the percentage - please review

**Comment by pingustar** [comment]
I have updated the design to be more aligned with the mockup. 

- I have removed the tabs
- moved the Claim section below the protected table
- updated the intial stake column
- updated the full coverage process bar

the latter is still work in progress. I need to do some more testing and cleanup.

@ashachaf current state is up on staging, if you wanna take a look.

**Comment by pingustar** [comment]
Progress bar and percentage is fixed.

@Velua ping me once we have the data to finish this off.

**Comment by ashachaf** [comment]
@pingustar 
updated tool tips:

Current Coverage (the column title for the counter coverage bar): 
The impermanent loss protection you have accrued. Impermanent loss protection starts 30 days after your deposit, at a rate of 30% and gradually increases 1% per day until you reach 100% protection.

Cliff: Impermanent loss protection starts 30 days after your deposit, at which point the "cliff" is reached and protection rises from 30%, by 1% per day, until 100% protection is reached.

**Comment by pingustar** [comment]
updated the tooltips.

**Comment by ashachaf** [comment]
@pingustar did you also add the tool tip to the column title?

**Comment by pingustar** [comment]
Yes, added column title tooltip for current coverage.

And updated cliff tooltip. 

Is it now showing on staging for you?

**Comment by ashachaf** [comment]
i see it now.

**Comment by ashachaf** [comment]
@pingustar 
in the "protected value" data,  we have a percentage increase indication.
this should represent the increase in value over the original stake.

calculating this:
(protected value / staked value -1) = percentage in decimals.
example:
(101.11/100-1)= 0.0111 -> turning to percentage, multiple this value by 100 -> 0.0111*100=1.11%.
(turning to ppm, multiple by 1000000 and trim the decimal part. example: 0.0111*1000000=11100

**Comment by ashachaf** [comment]
@Velua after checking this again, there is an updated calculation we should be using to extract the fees.

using the data you have extracted already:
Sqrt((opposingDepositedReserveCurrentBalance / depositedReserveCurrentBalance)/AddRate)*depositedReserveCurrentBalance/currentPoolTokenSupply* originalPoolTokenAmount -depositedAmount

due to rounding errors and differences with SMA rates, there is a workaround in the code that should also be supported in the value you display.

if (fees <0), display 0

**Comment by ashachaf** [comment]
protected value update:
given the indication above (average vs pool price), fully protected value should also change a bit.
calculate fully protected value.
check, if fully protected reserve amount < initial staked reserve amount, display initial staked reserve amount.

claimable value, 
calculate claimable value
check, if current time - initial stake time > maxProtectionDelay
  check, if claimable reserve amount < initially staked reserve amount, display initially staked reserve amount.

**Comment by ashachaf** [comment]
@pingustar 
tool tips are not aligned correctly (text to column).

please see correct text/column here (updated)

Tool tips:

Initial stake = amount of tokens you originally staked in the pool
Protected value = amount of tokens you can withdraw with 100% protection + fees
Claimable value = amount of tokens you can withdraw right now (assuming you have not earned full protection, this value will be lower than Protected Value)
Fees = fees your stake has earned since you entered the pool
ROI = the ROI of your fully protected value vs. your initial stake
[**updated**] APR = estimated calculation for yearly returns based on historical activity and data in the pool
[**updated**] Cliff: Impermanent loss protection starts immediately, but is available only once cliff is reached
Current Coverage (the column title for the counter coverage bar):
The impermanent loss protection you have accrued. Impermanent loss protection starts 30 days after your deposit, at a rate of 30% and gradually increases 1% per day until you reach 100% protection

**Comment by ashachaf** [comment]
@Velua fees calculation results:

      originalPoolTokenAmount: "1370909146698872748",
      currentPoolTokenSupply: "11338379267323634752139485",
      depositedAmount: "3874264344838181",
      depositedReserveCurrentBalance: "10932903826342227802656",
      opposingDepositedReserveCurrentBalance: "6847120160277936759163837",
      reserveRate: "290.61911993101308091657"
      // current result is -552144045218744

formula:
Sqrt((opposingDepositedReserveCurrentBalance / depositedReserveCurrentBalance)/reserveRate)*depositedReserveCurrentBalance/currentPoolTokenSupply* originalPoolTokenAmount -depositedAmount
Calculation:
Sqrt((6847120160277936759163837 / 10932903826342227802656)/290.61911993101308091657)*10932903826342227802656/11338379267323634752139485* 1370909146698872748 -3874264344838181
= - 0.0019337478 units in reserveToken

—

      originalPoolTokenAmount: "1370909146698872747",
      currentPoolTokenSupply: "11338379267323634752139485",
      depositedAmount: "1124989041279995704",
      depositedReserveCurrentBalance: "6847120160277936759163837",
      opposingDepositedReserveCurrentBalance: "10932903826342227802656",
      reserveRate: "0.00344092983365092823"
      // current result is -467584274913337187

formula:
Sqrt((opposingDepositedReserveCurrentBalance / depositedReserveCurrentBalance)/reserveRate)*depositedReserveCurrentBalance/currentPoolTokenSupply* originalPoolTokenAmount -depositedAmount
Calculation:
Sqrt((10932903826342227802656  / 6847120160277936759163837)/0.00344092983365092823)*6847120160277936759163837/11338379267323634752139485* 1370909146698872747 -1124989041279995704
= - 0.56103784 units in reserveToken

**Comment by ashachaf** [comment]
![image](https://user-images.githubusercontent.com/40430285/98912189-6b7ab980-24ce-11eb-8ed3-9e30b85e6130.png)

**Comment by ashachaf** [comment]
some USD value of the claimable value is larger than fully protected value.
example:
protected value: 129.27 tkn (~$73.05)
claimable value: 129.22 tkn (~$72.28) 
while tkn amount is lower, the usd value is higher.

![image](https://user-images.githubusercontent.com/40430285/98914157-fb216780-24d0-11eb-80ed-f6041981921c.png)

**Comment by Velua** [comment]
In staging there is a file ```pureHelpers.ts``` where we are doing the math for https://github.com/bancorprotocol/webapp/issues/426#issuecomment-717061509 this needs to be updated to reflect the new formula posted [here](https://github.com/bancorprotocol/webapp/issues/426#issuecomment-725351154)

Please write a test as Asaf has demonstrated in the file ```TestHelper.spec.ts``` so we know it's good to go. 

Asaf's new [comment](https://github.com/bancorprotocol/webapp/issues/426#issuecomment-725921595) may just have been fixed by 204899b2143f56f1fe92ea4c273f4f88df13cfbd

**Comment by ashachaf** [comment]
@pingustar 
in some cases the ROI is NaN%.
it might be related to the fact that the ROI is 0 or smaller.
if this is the case, please indicate 0.00%

![image](https://user-images.githubusercontent.com/40430285/98935390-21a0cc00-24ec-11eb-8f42-909dd3169770.png)

**Comment by pingustar** [comment]
fixed. ROI NaN issue. Added handling for < 0.01% amounts

**Comment by pingustar** [comment]
I have updated the formula. Result values are all below 0. Have tested with 6 positions. Still investigating. Fees column is hidden until then.

**Comment by pingustar** [comment]
@ashachaf closing this as this issue is 99% done and getting to big. Remaining "Fees column issue" moved to new task: #531
